tensorflow_quantum/python/differentiators/differentiator.py

```diff
@@ -179,6 +179,71 @@ def refresh(self):
             del self.expectation_op
         return self
 
+    @abc.abstractmethod
+    def get_gradient_circuits(self, programs, symbol_names, symbol_values):
+        """Return circuits to compute gradients for given forward pass circuits.
+
+        Prepares (but does not execute) all intermediate circuits needed to
+        calculate the gradients for the given forward pass circuits specified by
+        `programs`, `symbol_names`, and `symbol_values`. The returned
+        `tf.Tensor` objects give all necessary information to recreate the
+        internal logic of the differentiator.
+
+        If the caller has a list of functions `f` (one for each program in the
+        batch) with [n_vals] output values per function, then the derivative of
+        the computation will have shape [batch_size, n_params, n_vals]. In terms
+        of the Args and Returns of `get_gradient_circuits`, each entry of this
+        derivative can then be computed as
+
+        d(<programs[i](symbol_values[i])|
+            f[i]|programs[i](symbol_values[i])>)
+        / d(symbol_names[k]) = sum_m (batch_mapper[i, k, m] *
+            (<batch_programs[i, m](batch_symbol_values[i, m])|
+                f[i]|batch_programs[i, m](batch_symbol_values[i, m])>)
```

Comment:
This notation is kind of confusing. What exactly do you mean when you write `f[i] | batch_programs[i, m]` ?

Would it be easier to just explain: "This is how the returned values can be used to reconstruct the original gradients" ? I feel like with that explanation a user can just pick apart whichever component they want and develop a better understanding of what is happening. 

Thoughts ?

Reply:
`|batch_programs[i, m](batch_symbol_values[i, m])>` is the ket representing the state generated by program `batch_programs[i, m]` using the parameter values `batch_symbol_values[i, m]`.  Thus `<batch_programs[i, m](batch_symbol_values[i, m])|f[i]|batch_programs[i, m](batch_symbol_values[i, m])>` is the expectation value of function `f[i]` against that state.  I suppose if the notation is not clear enough here it won't be for other people either though.

Would the docstring be explicit enough for a user to combine the outputs appropriately if there were not an equation?  I'm deep into this enough that I can't quite tell, if it is I am fine to leave it out (this equation was meant to be the same as the one you commented on in the design doc).

An example of the list of vector-valued functions `f` is the PauliSums tensor used by the other ops.

Reply:
Oh also I realized that `f[i] | batch_programs[i, m]` starting a line is just due to line breaking for the 80 char restriction, if that is what makes it confusing I could stop the break and add a pylint ignore

Reply:
What about just writing a minimal running snippet that combines the outputs to compute a gradient ? Would that work ?

Reply:
Good point, that sounds better, I'll add a snippet that actually does it instead of writing up the math